jest-message-util rewrite using stack-tools

conartist6 opened the pull request:
See discussion on #11935.

conartist6 commented:
Why are snapshot files excluded from review by default? That seems like a way to end up with snapshots that assert results that no human would deep to be acceptable, and I found some of those when I was writing this.

conartist6 commented:
Prior to merging of this PR I will promote the `stack-tools` packages from `0.1.0` (I published them just so I could run the workflows here) to `1.0.0`. I believe they are already essentially ready, but I'd don't want to have to go to `2.0.0` if breaking API changes are necessary due to review changes.

conartist6 commented:
Any usage of `parseError` is suspect and still needs to be replaced with `parseErrors`.

conartist6 commented on e2e/native-async-mock/yarn.lock:
```diff
@@ -0,0 +1,11 @@
+# This file is generated by running "yarn install" inside your project.
+# Manual changes might be lost - proceed with caution!
+
+__metadata:
+  version: 4
+
+"root-workspace-0b6124@workspace:.":
+  version: 0.0.0-use.local
+  resolution: "root-workspace-0b6124@workspace:."
+  languageName: unknown
+  linkType: soft
```
Why did this get checked in? I believe the test suite generated it. I believe it should be .gitignored

conartist6 commented on packages/jest-circus/src/utils.ts:
```diff
@@ -332,20 +337,21 @@ export const makeSingleTestResult = (
     if (
       parsedLine &&
-      typeof parsedLine.column === 'number' &&
-      typeof parsedLine.line === 'number'
+      parsedLine.type === 'CallSiteFrame' &&
+      parsedLine.callSite.site.type === 'FileSite'
```
This is a good candidate for a sugar API, similar to the one I have provided for `getAbsoluteSitePath`. I still need to consider just how many such methods I want to provide and how.

conartist6 commented on packages/jest-circus/src/utils.ts:
```diff
@@ -419,6 +425,10 @@ const _getError = (
   return asyncError;
 };
 
+export const isError = (e: unknown): e is Error => {
+  return String(e) === '[object Error]' || e instanceof Error;
+};
+
```
I added this very early on and it may not longer be used anywhere. Consider removing.

conartist6 commented on packages/jest-message-util/src/__tests__/messages.test.ts:
```diff
@@ -27,63 +28,76 @@ const unixStackTrace =
   at Expectation.addResult (../jest-jasmine2/build/jasmine-2.4.1.js:591:21)
   at Expectation.toBe (../jest-jasmine2/build/jasmine-2.4.1.js:1504:12)
-  at Object.it (build/__tests__/messages-test.js:45:41)
+  at Object.it (/build/__tests__/messages-test.js:45:41)
```
`stack-tools` is a lot stricter than `stack-utils` was. As this test demonstrates certain changes are definitely breaking, though as far as I know they only break usages that do not occur in ~~the real world~~ modern versions of node.

conartist6 commented on packages/jest-message-util/src/__tests__/messages.test.ts:
```diff
@@ -27,63 +28,76 @@ const unixStackTrace =
-      at Object.asyncFn (node_modules/jest-jasmine2/build/jasmine_async.js:124:345)
-      at resolve (node_modules/jest-jasmine2/build/queue_runner.js:46:12)
+const assertionStack = stripIndent`
+  Expected value to be of type:
+    "number"
+  Received:
+    ""
+  type:
+    "string"
```
Jest's own test suite convinced me that it was necessary to preserve all whitespace in error messages, which means that good looking results require having input that is formatted in a reasonable way.

conartist6 commented on packages/jest-message-util/src/types.ts:
```diff
@@ -1,12 +1,5 @@
-/**
- * Copyright (c) Facebook, Inc. and its affiliates. All Rights Reserved.
- *
- * This source code is licensed under the MIT license found in the
- * LICENSE file in the root directory of this source tree.
- */
```
Oops I guess this is supposed to exist, though I don't know what the point is. I don't even think you could copyright this... Maybe there needs to be a lint rule making sure that these comments exist where they are supposed to?

conartist6 commented on packages/jest-message-util/src/visitor.ts:
```diff
@@ -0,0 +1,158 @@
+  indentHeader: string;
+  indentFrames: string;
+};
+
```
The `JestVisitor` class handles anything to do with the way jest formats any and all errors.

conartist6 commented on packages/jest-message-util/src/visitor.ts:
```diff
@@ -0,0 +1,158 @@
+    }
+    return printed;
+  }
+}
```
I probably want to export and use a `printNode(node)` function, just like I do in `stack-tools` itself.

conartist6 commented on packages/jest-message-util/src/visitor.ts:
```diff
@@ -0,0 +1,158 @@
+
+export class JestVisitor<O extends JestVisitorOptions> extends PrintVisitor<O> {
+  constructor(context: Context, options: O) {
+    super(context, options);
```
context is a stub really, and I haven't seen any evidence that I need it. I started with a purely functional design based off babel, but the class based system gave me a lot more safety, flexibility, and power. Since I have the visitor class prototype and instance for shared procedures and scratch data I think the correct thing to do is to eliminate the unused argument in stack-tools and here.

conartist6 commented on packages/jest-message-util/src/visitor.ts:
```diff
@@ -0,0 +1,158 @@
+  }
+
+  ErrorHeader(error: ErrorNode): string | undefined {
+    const {options} = this;
+    const name = this.visit(error.name);
+    const name_ = name === 'Error' ? undefined : name;
+    const message = error.message ? this.visit(error.message) : '';
+    const message_ = !name && message === 'Error' ? undefined : message;
+    const header =
+      message_ && name_ ? `${name_}: ${message_}` : name_ || message_;
+    return header && indentLines(header, options.indentHeader);
```
This logic is messy, but this is exactly the logic needed to get the tests to pass. At least it's all in one place where it is spelled out explicitly.

conartist6 reviewed (commented):
All in all I'm quite happy with how this worked out. The changes to jest itself are not too aggressive, and I hope they will open the door to more changes by making it clearer what the existing logic is.

conartist6 commented:
Looks like the build is busted on a single error, and it's one that I saw locally as well. I was really hoping it would go away in the sanctioned environment.
```
/home/runner/work/jest/jest/e2e/__tests__/detectOpenHandles.ts

  ● prints out info about open handlers

    expect(received).toMatchSnapshot()

    Snapshot name: `prints out info about open handlers 1`

    - Snapshot  - 1
    + Received  + 1

    @@ -1,8 +1,8 @@
      Jest has detected the following 1 open handle potentially keeping Jest from exiting:

    -   ●  TCPSERVERWRAP
    +   ●  GETADDRINFOREQWRAP

            12 | const app = new Server();
            13 |
          > 14 | app.listen({host: 'localhost', port: 0});
               |     ^

      58 |   const textAfterTest = getTextAfterTest(stderr);
      59 |
    > 60 |   expect(wrap(textAfterTest)).toMatchSnapshot();
         |                               ^
      61 | });
      62 |
      63 | it('does not report promises', () => {

      at Object.toMatchSnapshot (e2e/__tests__/detectOpenHandles.ts:60:31)
```

I don't get it. How could any of the code I wrote cause that?

conartist6 commented:
Some of the runs are also failing with:
```
/home/runner/work/jest/jest/e2e/__tests__/processExit.test.ts

  ● prints stack trace pointing to process.exit call

Error: expect(received).toMatchSnapshot()

Snapshot name: `prints stack trace pointing to process.exit call 1`

- Snapshot  - 11
+ Received  +  1

-   ●  process.exit called with "1"
-
-        6 |  */
-        7 |
-     >  8 | process.exit(1);
-          |         ^
-        9 |
-       10 | test('something', () => {
-       11 |   expect(true).toBe(true);
-
-       at Object.exit (__tests__/test.js:8:9)
+   ●      thrown: [Error: process.exit called with "1"]
    at Object.toMatchSnapshot (/home/runner/work/jest/jest/e2e/__tests__/processExit.test.ts:14:24)
    at Promise (/home/runner/work/jest/jest/packages/jest-circus/build/utils.js:396:28)
    at new Promise (<anonymous>)
    at callAsyncCircusFn (/home/runner/work/jest/jest/packages/jest-circus/build/utils.js:321:10)
    at _callCircusTest (/home/runner/work/jest/jest/packages/jest-circus/build/run.js:218:40)
    at process._tickCallback (internal/process/next_tick.js:68:7)
```

That's a conundrum. The code says:
```ts
const formatExecError = (
  error: Error | TestResult.SerializableError | string | undefined,
  ...
) => {
  // ...
}
```

There's no mention in those types about the possibility of error being an array. Also since there's only one element in the array that's breaking that test it's not obvious what the array is there for.

conartist6 commented on packages/jest-message-util/src/index.ts:
```diff
@@ -5,118 +5,86 @@
-    return warnAboutWrongTestEnvironment(error, 'jsdom');
-  } else if (error.includes('.unref is not a function')) {
-    return warnAboutWrongTestEnvironment(error, 'node');
+// export const stripIndent = (lines: string): string => {
+//   const match = lines.match(/^[^\S\n]*(?=\S)/gm);
+//   const indent = match && Math.min(...match.map(el => el.length));
+//   if (indent) {
+//     const regexp = new RegExp(`^.{${indent}}`, 'gm');
+//     return lines.replace(regexp, '');
+//   }
+//   return lines;
+// };
```
oops

conartist6 commented on package.json:
```diff
@@ -155,6 +156,7 @@
     "jest-environment-node": "workspace:*",
-    "react-native": "patch:react-native@0.64.0#./patches/react-native.patch"
+    "react-native": "patch:react-native@0.64.0#./patches/react-native.patch",
+    "@jest/monorepo": "portal:/home/conartist6/dev/repos/jest"
```
Why did yarn do this? Every time I delete it it keeps putting it back.

conartist6 commented on packages/jest-message-util/src/types.ts:
```diff
@@ -1,12 +1,5 @@
-
-import type {StackData} from 'stack-utils';
-
-export interface Frame extends StackData {
```
This is also technically breaking since this type used to be a good bit wider. I could fix it so that it includes all the data it used to relatively easily though

conartist6 commented on packages/jest-message-util/src/index.ts:
```diff
@@ -5,323 +5,294 @@
-const ANONYMOUS_PROMISE_IGNORE = /^\s+at (new )?Promise \(<anonymous>\).*$/;
-const ANONYMOUS_GENERATOR_IGNORE = /^\s+at Generator.next \(<anonymous>\).*$/;
-const NATIVE_NEXT_IGNORE = /^\s+at next \(native\).*$/;
-const TITLE_INDENT = '  ';
-const MESSAGE_INDENT = '    ';
-const STACK_INDENT = '      ';
```
I removed these since their usage is inconsistent, i.e. message is not always indented four spaces and not everything indented four spaces is a message. There are many, many ways to figure out how many spaces there are in a string.

conartist6 commented on packages/jest-message-util/src/index.ts:
```diff
@@ -5,323 +5,294 @@
+  options: StackTraceOptions = {},
+): Array<FrameNode> => {
+  let frames_ = frames || [];
+  frames_ = frames_.filter(frame => !isInternalFrame(frame));
+
+  const firstFrame = frames_[0];
+
+  frames_ = frames_.filter(frame => !isJestInternalFrame(frame));
```
This double filtering looks a little odd, but I prefer it since it means the callback can be completely stateless, where the previous callback needed to check whether it had ever previously returned true.

conartist6 commented:
The file handles bug stems from a fundamental ambiguity in node errors. Errors have a `message` and a `stack`, but the first line of the stack is (usually) the content of `error.message`. If you write to `error.message` of course, then you've caused these two sources of truth to go out of sync.

To complicate things, changing an error's message may be just fine -- so long as you've never tried to print the error or in any other way access `error.stack` (e.g. in a debugging session). This is because the content of `error.stack` is generated lazily by node the first time it is accessed. If you write to `error.message` before reading `error.stack`, then you'll stay in sync.

Finally this problem has already been anticipated to some degree. I added a `replaceMessage` method to `stack-tools` which ensures that `error.stack` is updated, and of course another way to extend errors is using the `error.cause` property which hopefully in the longer run will alleviate the need to tamper with error messages.

All that said, partly because this issue can be so finnicky (e.g. affected by debugging), I think my current approach could create some real pain. I use both sources of truth in different places, which is to say I assume that they are in sync. `stack-tools` does its basic parsing by eliminating `error.message` from the top of `error.stack`, meaning it treats `error.name` and `error.message` as the source of truth. `@stack-tools/node-tools` meanwhile always parses `error.stack`, deducing `error.name` and `error.message` from them while ignoring their real values.

It seems reasonable to assume that if `error.message` is different than `error.stack` then `error.message` is likely the real source of truth as to that value. Since `@stack-tools/v8-tools` is able to parse a stack no matter what header is on top, it seems like the smart money would be to just overwrite the values parsed from `stack` with the `name` and `message` property values when the object being parsed is an error.

conartist6 commented:
I fixed several more issues including the message/stack ambiguity mentioned above, though that did lead to a few more errors surfacing. I'm currently seeing a failure in `e2e/__tests__/unexpectedToken.test.ts` and I don't know why, because I'm quite sure it was passing just recently and the output certainly looks like it's working as intended. Anyway, I can dig into that tomorrow.

conartist6 commented:
These last errors seem to be caused by some textual (that is to say non-frame) error content which is in the `stack` but not the `message`. This is likely something babel itself is doing so there's no changing it. Should be easy enough to handle though I'd think.

conartist6 commented:
That also explains why it broke. Before I was parsing message from stack so everything in stack was definitely there. Now I'm getting everything but frames from `error.message` which caused this problem.

mrazauskas commented on packages/jest-core/src/collectHandles.ts:
```diff
@@ -9,10 +9,10 @@
 
 import * as asyncHooks from 'async_hooks';
 import {promisify} from 'util';
-import stripAnsi = require('strip-ansi');
```
It seems like `strip-ansi` could be removed from `jest-core/package.json` too. Or?

conartist6 commented:
The other interesting thing is that the current discrepancy in message/stack consists of the code frame showing the syntax error.
```
/tmp/unexpected-token/node_modules/untranspiled-module:1
({"Object.<anonymous>":function(module,exports,require,__dirname,__filename,jest){import {module} from "some-module"
                                                                                  ^^^^^^
```

It's interesting because this code frame actually isn't that good. It may be quite useful that it's being elided. We could instead replace it with an `@babel/code-frame` like we use everywhere else in jest, and even better it might be possible to have the frame show the real content of `untranspiled-module` which are just `import {module} from "some-module"`.

conartist6 commented:
OK the reality is a little weirder here. In part it's weird because as far as I can tell the Node APIs are wrong (!). The problem here is that the docs for the feature that's mucking with the error indicate that it's part of the behavior of [script.runInContext](https://nodejs.org/api/vm.html#scriptrunincontextcontextifiedobject-options), where the behavior would be documented in and controlled by the `displayErrors` option. Unfortunately it seems that parsing of the code is done at the time of construction of the `Script` object (which does not take that option) making it completely useless.

So node is mucking with the stack traces, nobody can do anything about it, it attaches ugly code frames that are essentially wrong, and furthermore it provides none of the information necessary to construct a better message without attempting to parse the first one. Not too impressed with you today node, not too impressed.

So yeah the only way to get correct results is to make sure stack-tools can handle this.

conartist6 commented:
I think it's worth a try to see if that solution can be made to work well. Of course I'll also have to consider what to do in `parseError`, which never returns an `ErrorChain` node.

The nuclear option would be to eliminate it completely and just say that the calling code must always handle the fact that any caught error may actually be a chain of related errors. That might actually be pretty useful, since it would allow me to use the convenience of the code I've built to push developers to always handle chains of errors. Anyone who needed a single error could do `parseErrors(error)[0]`, which I kind of like because I don't want people doing that -- it's potentially discarding information about what when wrong, which is the kind of thing that should be as evident as possible and written in a place that a casual reader can discover it. Tucking the possibility of a discarding error information away behind a polished API in a widely used library seems to a degree to be asking for the possibility that lazy programmers never really adopt `error.cause` at all, instead propagating a quiet assumption that errors are single entities and not related chains.

conartist6 commented:
Of course I should consider that I'm really talking about two use cases: one which fits this narrative and one which does not. Node is just tossing a code frame onto the error, which isn't really an error of its own. I should actually consider making code frames a first-class part of `stack-tools` as they are shared functionality which I think many people will be interested in. In that case parsing a code frame into its own error is wrong. But in this same test jest is prepending a big explanatory block onto that error, and that definitely could be better expressed with cause... which I can just do!

conartist6 commented:
That code frame isn't really the same as other code frames we print though. It's not conceptually part of the stack since it details what happened, not where it happened.

conartist6 commented:
The absolute hackiest hack would just be to dictate that `` `${name}: ${message}` `` is always printed on top of an error. Anything that was prefixed to it becomes suffixed to it and tough beans if you don't like it.

Given how hacky that solution is, I actually think I rather like it. It doesn't discard any information, it doesn't rely on really understanding or parsing any of the content, and it's pretty easy to explain what's being done and why.

conartist6 commented:
It's tough partly because the whole thing is so wishy-washy. Any heuristic I come up with to support one case breaks another. I basically just have to admit that some cases are going to be broken, and ensure that if the tool I come up with succeeds that people are able to work around any problems I am creating for them.

For example with the relocation heuristic I'd be breaking a pattern that current version of Jest use, specifically prepending a complete additional message to stack like:
```
error.stack = `Jest has encountered ...
...
Details:
` + error.stack;
```

Now essentially this is creating an error chain by hand, so I think I'm fine with breaking that because I'm already offering the better workaround or constructing an actual chain. That said, that workaround will only work for users who are environments which will actually print error chains, which currently more or less 0% of environments.

conartist6 commented:
There is a workaround for this problem though. If you want to create an error chain but you don't know if your environment will print data stored in `error.cause` you can just alter the error in such a way as to ensure that parsing it would produce multiple errors.

Right now that means supplying an omitted stack frame, i.e. `at <omitted>`. Now that I think about this I think I've implemented it wrong. First, `<omitted>` means that the entire stack is omitted, and yet I've created a node for it at the frame level. This requires additional checks that could be unnecessary, e.g. `frame.type !== 'OmittedFrame'`. Second, because the syntax needed is `at <omitted>`, it a) reads wrong and b) can't be used if the format of the stack trace is not known, which may be the case if the user is only using the base `stack-tools` package for its runtime methods like `replaceMessage(error)`.

Instead I think I should have a syntax that is parsed at the error level that can be used to denote a break in the cause chain (in lieu of a stack, which also does that). A small tweak might be to use something like `<frames omitted>`, but I'm not so sure about that. I had initially assumed that only `stack-tools` itself would generate such a thing, but now that I'm considering the possibility that users would be writing and reading this notation possibly with some frequency I might want something more polished. I'm going to transition this discussion over to stack-tools itself since it really belongs there.

conartist6 commented:
Further discussion on this thread is on https://github.com/stack-tools-js/stack-tools/issues/15. I've rewritten a bunch of the code in a way that I think greatly improves the consistency of experience. The base `stack-tools` package is now able to parse strings and error chains, it just always returns `TextFrame` nodes, which now no longer include the `at ` prefix in `frame.text`. The `@` symbol is now considered an alternative to `at ` as a frame prefix. Both parse equally well, and a new `glyphs` option has been introduced to control the printing behavior. I've also removed `OmittedFrame` from the AST.

I've accomplished these improvements with a new no-lib error parser.

SimenB commented on packages/jest-message-util/src/visitor.ts:
```diff
@@ -0,0 +1,165 @@
+
+const PATH_NODE_MODULES = `/node_modules/`;
+const PATH_JEST_PACKAGES = `/jest/packages/`;
+const STACK_TRACE_COLOR = chalk.dim;
```
I'd inline this

SimenB commented on packages/jest-message-util/src/visitor.ts:
```diff
@@ -0,0 +1,165 @@
+    const {options} = this;
+    const name = this.visit(error.name);
+    const name_ = name === 'Error' ? undefined : name;
+    const message = error.message ? this.visit(error.message) : '';
+    const message_ = !name && message === 'Error' ? undefined : message;
```
```suggestion
    let message = error.message ? this.visit(error.message) : '';
    message = !name && message === 'Error' ? undefined : message;
```

SimenB commented on packages/jest-message-util/src/visitor.ts:
```diff
@@ -0,0 +1,165 @@
+
+  ErrorHeader(error: ErrorNode): string | undefined {
+    const {options} = this;
+    const name = this.visit(error.name);
+    const name_ = name === 'Error' ? undefined : name;
```
```suggestion
    let name = this.visit(error.name);
    name = name === 'Error' ? undefined : name;
```

SimenB commented on packages/jest-runner/src/testWorker.ts:
```diff
@@ -6,6 +6,7 @@
  *
  */
 
+import {parseError, printError, printNode} from '@stack-tools/node-tools';
```
thoughts on exporting some helpers from `jest-message-util` for this? A single `separateMessageFromStack` is easier than using these 3 functions, and keeping that complexity inside `jest-message-util` (like today) seems advantageous to me

SimenB commented on packages/jest-runtime/src/index.ts:
```diff
@@ -18,6 +18,7 @@ import {
   // @ts-expect-error: experimental, not added to the types
   Module as VMModule,
 } from 'vm';
+import {getAbsoluteSitePath, parseError} from '@stack-tools/node-tools';
```
same here, would prefer to have the logic in `jest-message-util`

SimenB reviewed (changes requested):
exciting stuff! Since it's in draft I just skimmed through, but this is looking promising 🙂

conartist6 commented:
Also just because I have started so much ambitious work and burned out on it, I've deprioritized this. This was a trip down the rabbit hole from the build system I set out to create, and in order to return myself to a semblance of a productive being it's that project I need to finish to break the logjam. So it's likely that this work will sit a while longer

SimenB commented:
Totally understandable, this is a herculean task 😀 Get well, and come back stronger whenever you feel ready (or don't, none of us are getting paid for this 🙂)

conartist6 commented:
Just saw another syntax I've never seen before which I observed when jest caught an error thrown by `spawnSync`
```
      at node:internal/main/run_main_module:17:47 {
        code: 'MODULE_NOT_FOUND',
        requireStack: []
      }
      at run (lib/utils/shell.js:11:15)
      at Object.<anonymous> (test/test-project.js:20:5)
 ```

mrazauskas commented on packages/jest-message-util/src/index.ts:
```diff
@@ -5,323 +5,296 @@
+// `before/after each` hooks). If it's thrown, none of the tests in the file
+// are executed.
+export const formatExecError = (
+  error: Error | TestResult.SerializableError | string | undefined,
```
It might make sense to have `error: unknown` here. In a way `number`, `null`, or some object would be handled. Type could allow these. I was wrestling with this idea for some time till my head remembered this PR. Just leaving a note.

GerkinDev commented:
Hi @conartist6 , need a hand on this ? I'd be glad to have this behavior merged: I quite heavily rely on `error.cause` to give context to errors, so not having the underlying cause is a real problem for me.

conartist6 commented:
@GerkinDev I am full time open source working on different projects. This PR sank because my stack-tools library sank. It's simply not possible to build what I was trying to build. The only way you can understand anything about stack traces is if you're absolutely sure they're node stack traces, and you're still going to be bitten by the fact that their formatting can change at any time and in fact it has changed significantly with the most modern versions of node since I first wrote this PR.

If you'd like the functionality I recommend you start over with a new PR that salvages as much from this one as make sense.
